### ISSUE TITLE
CKEditor: Make images work when Backdrop is installed in a subdirectory of the server root.

### ISSUE THREAD

**Issue by SamStamport** [backdrop/backdrop-issues#2560]
When I click the Image button at the top of CKEditor the HTML generated is `<img alt="" data-align="center" data-file-id="335" src="/sites/default/files/inline-images/IMG_20151013_180713.jpg" />`. This does not result in the proper display of the actual image. I get a little image icon instead. What am I doing wrong?

**Comment by olafgrabienski** [comment]
I can't reproduce the behaviour on a test site. @ColumbiaBlooms, can you inspect the image element in the browser and post the HTML code? Here an example of my working site:

`<img alt="Dummy image" data-file-id="1" src="/files/inline-images/dummy_300.png" height="200" width="300">`

**Comment by ColumbiaBlooms** [comment]
Here's the rendered (is that the right term?) HTML ...

<img alt="" data-file-id="335" src="/sites/default/files/inline-images/IMG_20151013_180713.jpg" class="align-center">

When I edit the article and turn on source this is what I see 

<img alt="" data-align="center" data-file-id="335" src="/sites/default/files/inline-images/IMG_20151013_180713.jpg" />

It seems odd that the upload process stored this image in /sites/default/files instead of just /files/.

I'm using Backdrop 1.6.0 running under XAMPP with Windows 10. This problem seems to be related to https://github.com/backdrop/backdrop-issues/issues/2261#issuecomment-271403890 & https://github.com/backdrop/backdrop-issues/issues/2220.

**Comment by olafgrabienski** [comment]
@ColumbiaBlooms - One 'technical' hint: to display a HTML tag in Github comments, you have to mark it as code. You can use the `< >` button in the comment's *Write* toolbar to do so and get the following result for the rendered HTML:

`<img alt="" data-file-id="335" src="/sites/default/files/inline-images/IMG_20151013_180713.jpg" class="align-center">`

The path `/sites/default/files` looks like a Drupal path to me. Is it possible that your Backdrop site has inherited some configuration from a former Drupal site?

**Comment by ColumbiaBlooms** [comment]
Yes. There may be debris left over from when I ported my D7 site to Backdrop. How can I clean this up?

**Comment by klonos** [comment]
I have yet to confirm this by doing an upgrade of a stock D7 site + a WYSIWYG module to latest Backdrop, but it seems that this happens only after upgrading a site from D7, so changing the issue title accordingly.

**Comment by ColumbiaBlooms** [comment]
As an experiment I edited the article, turned on source, and changed the link in question to ` src="/files/IMG_20151013_180713.jpg"`. The effect was the same. The image was not rendered.

**Comment by olafgrabienski** [comment]
Does the issue happen in a local environment or on a remote site, or both?

**Comment by ColumbiaBlooms** [comment]
Sorry. Can't answer that question. I only have Backdrop running locally due to this problem and others.

I could give you a copy of my file structure & database for the D7 site. Would that help?

**Comment by Graham-72** [comment]
@ColumbiaBlooms what is your setting for Public File System Path at /admin/config/media/file-system?
I have a site imported from D7 and my setting is sites/default/files because that is where all the files were for the D7 site. New uploaded files are put there too.

**Comment by olafgrabienski** [comment]
@ColumbiaBlooms Thanks, no need of a copy. Is the site URL just a local domain or subdomain (like *mysite.dev* or *backdrop.mysite.dev*) or does it contain an appended folder name (like *mysite.dev/backdrop*)?

If the latter, you could try something like `src="/backdrop/files/IMG...jpg"` instead of `src="/files/IMG....jpg"`. (If that works, migrating the site from local to a remote site could get difficult, but that's another issue.)

**Comment by ColumbiaBlooms** [comment]
@Graham-72 Ah. Yes. The Public File System Path solves the problem of where uploaded files are placed. I fixed mine. Would a code update for the migration from D7 fix this problem for others or can a documentation update be made so that others know to fix this?

@olafgrabienski - My local site is stored at C:\xampp\htdocs\BD. Changing the link to `src="/bd/files/IMG_20151013_180713.jpg` solves the image display problem locally. Pathologic seems to have been ported to Backdrop and is fully functional. If I install it will it solve the problem of migrating to my live site with `src="/bd/files/IMG_20151013_180713.jpg` coded this way?

**Comment by Graham-72** [comment]
@ColumbiaBlooms I would suggest that it is perfectly valid to continue to use sites/default/files as your Public File System Path and it would not be appropriate to change this automatically when migrating from Drupal, the reason being that is where all the previously uploaded files reside.

I often install my sites in a subdirectory of htdocs (or in my case httpdocs) and this is why I need to use settings and .htaccess as described in my comment https://github.com/backdrop/backdrop-issues/issues/2220#issuecomment-271938135.  Does XAMPP provide for use of redirects in .htaccess? It is a few years since I had a local site running in Windows.

**Comment by klonos** [comment]
> I have a site imported from D7 and my setting is sites/default/files because that is where all the files were for the D7 site. New uploaded files are put there too.

> Ah. Yes. The Public File System Path solves the problem of where uploaded files are placed.

Never tried an actual upgrade from D7, but according to our "Upgrading from Drupal 7" documentation, this shouldn't have happened:

> 2. Copy the files directory from Drupal 7 (usually drupal/sites/default/files) into the Backdrop site (usually backdrop/files)

@Graham-72 did you follow or (intentionally) skip this step and ended up with your files still under `/sites/default/files`?

@ColumbiaBlooms from what you are saying, I'm assuming that you did follow this step of moving those files to `/files`. Are you saying that after the upgrade, the path was not changed in the file system settings? If so, then I believe that this is something that needs to be fixed.

> Would a code update for the migration from D7 fix this problem for others or can a documentation update be made so that others know to fix this?

I would expect things to "just work". @ColumbiaBlooms in the original issue description, you say that the automatically generated `src` path (`/sites/default/files/...`) does not work and images are not displayed. If you have moved your files to `/files`, then that makes sense. If you haven't move your files from ``/sites/default/files`, then (theoretically) it should work.

**Comment by ColumbiaBlooms** [comment]
@Graham-72 I'm glad that changing settings.php & .htaccess worked for you, but I'd rather not have to fool around with modifying settings.php & .htaccess. A major reason I like Backdrop is that I don't have to mess with these files. I also like the cleaner & much simplified file structure in Backdrop.

@klonos Yes. I followed the instructions and copied my D7 files to `/files`. It does appear that the path was not automatically changed as part of the migrate process from D7 to Backdrop. I just now did another test using the image button. `src="/files/inline-images/Chrysanthemum.jpg"` is what was generated even though Public file system path is set to "files". Where is this "inline-images" coming from?

**Comment by klonos** [comment]
> Yes. I followed the instructions and copied my D7 files to `/files`. It does appear that the path was not automatically changed as part of the migrate process from D7 to Backdrop.

...and that's precisely what I thought that the issue is. We need to either:

1. Remove that step from documentation or tell people that it is optional and that it is fine if they leave their files in the old `/sites/default/files` location
2. Make it so that during a migration the file system settings are adjusted accordingly ...and perhaps (??) do the copying/moving of the files in the proper `/files` directory either automatically or after user confirmation. In other words, checkbox set with two options during installation:
  - move files from `/sites/default/files` to `/files`  <-- sets the public file system path to `/files`
  - leave files under the old `/sites/default/files` location"  <-- sets the public file system path to `/sites/default/files`

...and either way, we need to make sure that inserting an image still works.

> Where is this "inline-images" coming from?

I believe that this is a subfolder that is automatically created in order to hold all images inserted via the WYSIWYG editor.

**Comment by ColumbiaBlooms** [comment]
I love the simplified file structure in Backdrop compared to Drupal. I'd vote to NOT give an option to put files in `/sites/default/files`. It makes finding files less intuitive and makes the file structure more complicated.

So, it seems that clicking the image button still does not work locally. What works is `src="/bd/files/IMG_20151013_180713.jpg`, but the image button generates `src="/files/inline-images/Chrysanthemum.jpg"`. Or maybe I'm just confused ...

**Comment by ColumbiaBlooms** [comment]
Why would installing in a subdirectory be a problem? A directory is a directory is it not? I guess I'm naive, but it seems to me that there must be a way to resolve this with some code so that the image button just works resulting in an src that works locally and (possibly with Pathologic installed) works when porting to the live site without having to manually change a lot of links.

**Comment by olafgrabienski** [comment]
@ColumbiaBlooms It shouldn't be a problem, but it is. (Actually, the problem isn't to install Backdrop in a subdirectory but to use the subdirectory as part of the site URL like `example.com/backdrop`.)

The problem seems to have a long history, and as far as I know at the moment there are only workarounds. I recommend to read https://github.com/backdrop/backdrop-issues/issues/2220#issuecomment-249192177 and the following posts to get a picture.

PS: don't know if we should tag #2220 as a bug to get the "subdirectory problem" solved (the issue is, like this one, also about other questions), or open a new issue for it. @klonos - what do you think?

**Comment by Graham-72** [comment]
The problem, if I may call it that, stems from the change introduced for good reason in issue #2261. 
I commented there on the difficulty this caused with subdirectories, see https://github.com/backdrop/backdrop-issues/issues/2261#issuecomment-271403890. 

I think the intention was that CKEditor should provide links relative to the base path, which would surely include the subdirectory would it not e.g mydomain.com/backdrop/? But it does not seem to be using the full base path including subdirectory when rendering the html. However, I may have misunderstood this.

**Comment by ColumbiaBlooms** [comment]
Hmm ... This does, indeed, seem like a complicated problem. From reading 2220 & 2261 it looks like the problem was solved, then not solved. References are made to this problem with blocks, but my problem is with a plain old article content type.

I wish I knew enough about PHP & the innards of Drupal/Backdrop to help, but I don't. I'm hoping that by opening this problem that a solution can be found and implemented soon. This seems to be basic functionality that should just work.

**Comment by docwilmot** [comment]
@ColumbiaBlooms there are two things going on here:
1. You've installed your test site into a subdirectory, so when you insert images via CKEditor, they are getting a path that doesnt match your file structure. Unfortunately, this isn't a simple fix. If you are going to build your site and plan to move it to another server, you really must have your test site in a subdomain if you want the images to work, otherwise the images are likely to have the wrong paths attached. Thats just the way it is for now, and I dont think its eay to fix, as mentioned. I've had this issue, I've checked.
2. Backdrop uploads all images inserted via the CKEditor to a folder in your "public" directory (`/files` usually) to a subfolder called `inline-images`. So really if your subdomain issue was sorted, your images would actually be in the right place.

**Comment by Graham-72** [comment]
@ColumbiaBlooms I would be interested to know what value you have for $base_dir in your installation? I think we now have CKEditor doing the 'right' thing in referencing relative to the site URL. I believe that rendering this is handled in core by function file_create_url($uri) in file.inc, which prepends $base_url. Have you set a value for $base_url in settings.php (line 91 or thereabouts)?

My complicated solution using .htaccess (https://github.com/backdrop/backdrop-issues/issues/2220#issuecomment-271938135) deals with a more complex situation where the installation is in a subdirectory but this is not evident to the user i.e. the site root appears to be at www.mydomain.com but is in fact at www.mydomain.com.backdrop

I wonder whether core function file_create_url($uri) could be modified to recognise automatically that the installation is not at the site root but is in a subdirectory?

**Comment by ColumbiaBlooms** [comment]
Thanks @docwilmot & @Graham-72. My $base_url in settings.php is commented out. When I start my site in Chrome I type http://localhost:8080/bd/. "bd" is the directory that contains my Backdrop site under XAMPP. (I have other Drupals such as my D7 site in other directories which makes maintenance easy.) 

I uncommented $base_url and set it to 'http://localhost:8080/bd/'. Public file system path is set to "files". When I insert an image I get src="/files/inline-images/p8_0.jpg" which does not work. I guess this does not work because it is not a subdomain? Please let me know if this is the reason. It looks like creating a subdomain in XAMPP is not a really big deal. I found https://www.techlila.com/how-to-create-subdomains-using-xampp/ that seems fairly simple. 

This business of inserting images is making my brain tired!

**Comment by ColumbiaBlooms** [comment]
Sorry hit the close button my mistake .. 'my bad!

**Comment by ColumbiaBlooms** [comment]
I've discovered that setting up a subdomain in XAMPP is a really big deal. I con't seem to get it to work. I keep getting page not found errors. I'm at a loss to figure out what to do next.

**Comment by herbdool** [comment]
@ColumbiaBlooms you could try Kalabox.io or set up a test site on a host where you can cheaply (or for free) create a temporary domain for a dev site.

**Comment by ColumbiaBlooms** [comment]
I don't quite understand why Kalabox.io would solve this image embed problem in CKEditor. Could you elaborate?

**Comment by klonos** [comment]
@ColumbiaBlooms I believe that with each Kalabox instance you get these:

```
http://mydrupalsite1.local
http://myotherdrupalsite.local
http://mywordpresssite.local
http://mybackdropsite.local
```

...instead of these:

```
http://localhost:8080/mydrupalsite1
http://localhost:8080/myotherdrupalsite
http://localhost:8080/mywordpresssite
http://localhost:8080/mybackdropsite
```

**Comment by ColumbiaBlooms** [comment]
Unable to install Kalabox. I filed a report at https://github.com/kalabox/kalabox/issues/1791. Waiting for answer ...

**Comment by ColumbiaBlooms** [comment]
Update: No response to https://github.com/kalabox/kalabox/issues/1791 so far. I found https://github.com/docker/for-win/issues/412 which is identical to my problem, but there's no resolution. 

Seems the only option I have left is @Graham-72's .htaccess solution. I'll try it and post results here.

**Comment by ColumbiaBlooms** [comment]
@The name of my subdirectory is "bd". It is at C:\xampp\htdocs. I inserted ``RewriteRule ^(.*)$ bd/$1 [L] (at line 55 approx.). I guess I misunderstood the procedure because I'm getting a couple of AH00124: Request exceeded the limit of 10 internal redirects due to probable configuration error. Use 'LimitInternalRecursion' to increase the limit if necessary. Use 'LogLevel debug' to get a backtrace. error messages in the log. 

What did I do wrong? Maybe I put the line in the wrong place?

Contents of .htaccess: 

`#
# Apache/PHP/Backdrop settings:
#

# Protect files and directories from prying eyes.
<FilesMatch "\.(engine|inc|info|install|make|module|profile|test|po|sh|.*sql|theme|tpl(\.php)?|xtmpl)(~|\.sw[op]|\.bak|\.orig|\.save)?$|^(\..*|Entries.*|Repository|Root|Tag|Template)$|^#.*#$|\.php(~|\.sw[op]|\.bak|\.orig\.save)$">  
  <IfModule mod_authz_core.c>
    Require all denied
  </IfModule>
  <IfModule !mod_authz_core.c>
    Order allow,deny
  </IfModule>
</FilesMatch>

# Don't show directory listings for URLs which map to a directory.
Options -Indexes

# Make Backdrop handle any 404 errors.
ErrorDocument 404 /index.php

# Set the default handler.
DirectoryIndex index.php index.html index.htm

# Override PHP settings that cannot be changed at runtime. See
# settings.php and backdrop_environment_initialize() in includes/bootstrap.inc
# for settings that can be changed at runtime.
<IfModule mod_php5.c>
  php_flag session.auto_start               off
  php_value mbstring.http_input             pass
  php_value mbstring.http_output            pass
  php_value max_input_vars                  10007
  php_flag mbstring.encoding_translation    off
</IfModule>

# Requires mod_expires to be enabled.
<IfModule mod_expires.c>
  # Enable expirations.
  ExpiresActive On

  # Cache all files for 2 weeks after access (A).
  ExpiresDefault A1209600

  <FilesMatch \.php$>
    # Do not allow PHP scripts to be cached unless they explicitly send cache
    # headers themselves. Otherwise all scripts would have to overwrite the
    # headers set by mod_expires if they want another caching behavior. This may
    # fail if an error occurs early in the bootstrap process, and it may cause
    # problems if a non-Backdrop PHP file is installed in a subdirectory.
    ExpiresActive Off
  </FilesMatch>
</IfModule>

# Various rewrite rules.
<IfModule mod_rewrite.c>
  RewriteEngine on

    # fix CKEditor embedded image URL problem 
    # This rewrite rule enables the /files/inline-images folder to be correctly found and used.
    RewriteRule ^(.*)$ bd/$1 [L]

  # Set "protossl" to "s" if we were accessed via https://.  This is used later
  # if you enable "www." stripping or enforcement, in order to ensure that
  # you don't bounce between http and https.
  RewriteRule ^ - [E=protossl]
  RewriteCond %{HTTPS} on
  RewriteRule ^ - [E=protossl:s]

  # Make sure Authorization HTTP header is available to PHP
  # even when running as CGI or FastCGI.
  RewriteRule ^ - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]

  # Block access to "hidden" directories whose names begin with a period. This
  # includes directories used by version control systems such as Subversion or
  # Git to store control files. Files whose names begin with a period, as well
  # as the control files used by CVS, are protected by the FilesMatch directive
  # above.
  #
  # The ".well-known" directory used by Let's Encrypt is access white-listed.
  #
  # NOTE: This only works when mod_rewrite is loaded. Without mod_rewrite, it is
  # not possible to block access to entire directories from .htaccess because
  # <DirectoryMatch> is not allowed here.
  #
  # If you do not have mod_rewrite installed, you should remove these
  # directories from your webroot or otherwise protect them from being
  # downloaded.
  RewriteCond %{REQUEST_URI} !^/\.well-known
  RewriteRule "(^|/)\." - [F]

  # If your site can be accessed both with and without the 'www.' prefix, you
  # can use one of the following settings to redirect users to your preferred
  # URL, either WITH or WITHOUT the 'www.' prefix. Choose ONLY one option:
  #
  # To redirect all users to access the site WITH the 'www.' prefix,
  # (http://example.com/... will be redirected to http://www.example.com/...)
  # uncomment the following:
  # RewriteCond %{HTTP_HOST} .
  # RewriteCond %{HTTP_HOST} !^www\. [NC]
  # RewriteRule ^ http%{ENV:protossl}://www.%{HTTP_HOST}%{REQUEST_URI} [L,R=301]
  #
  # To redirect all users to access the site WITHOUT the 'www.' prefix,
  # (http://www.example.com/... will be redirected to http://example.com/...)
  # uncomment the following:
  # RewriteCond %{HTTP_HOST} ^www\.(.+)$ [NC]
  # RewriteRule ^ http%{ENV:protossl}://%1%{REQUEST_URI} [L,R=301]

  # Modify the RewriteBase if you are using Backdrop in a subdirectory or in a
  # VirtualDocumentRoot and the rewrite rules are not working properly.
  # For example if your site is at http://example.com/backdrop uncomment and
  # modify the following line:
  # RewriteBase /backdrop
  #
  # If your site is running in a VirtualDocumentRoot at http://example.com/,
  # uncomment the following line:
  # RewriteBase /

  # Redirect common PHP files to their new locations.
  RewriteCond %{REQUEST_URI} ^(.*)?/(update.php) [OR]
  RewriteCond %{REQUEST_URI} ^(.*)?/(install.php) [OR]
  RewriteCond %{REQUEST_URI} ^(.*)?/(cron.php)
  RewriteCond %{REQUEST_URI} !core
  RewriteRule ^ %1/core/%2 [L,QSA,R=301]

  # Pass all requests not referring directly to files in the filesystem to
  # index.php. Clean URLs are handled in backdrop_environment_initialize().
  RewriteCond %{REQUEST_FILENAME} !-f
  RewriteCond %{REQUEST_FILENAME} !-d
  RewriteCond %{REQUEST_URI} !=/favicon.ico
  RewriteRule ^ index.php [L]

  # Rules to correctly serve gzip compressed CSS and JS files.
  # Requires both mod_rewrite and mod_headers to be enabled.
  <IfModule mod_headers.c>
    # Disable content sniffing, since it's an attack vector.
    Header set X-Content-Type-Options nosniff

    # Serve gzip compressed CSS files if they exist and the client accepts gzip.
    RewriteCond %{HTTP:Accept-encoding} gzip
    RewriteCond %{REQUEST_FILENAME}\.gz -s
    RewriteRule ^(.*)\.css $1\.css\.gz [QSA]

    # Serve gzip compressed JS files if they exist and the client accepts gzip.
    RewriteCond %{HTTP:Accept-encoding} gzip
    RewriteCond %{REQUEST_FILENAME}\.gz -s
    RewriteRule ^(.*)\.js $1\.js\.gz [QSA]

    # Serve correct content types, and prevent mod_deflate double gzip.
    RewriteRule \.css\.gz$ - [T=text/css,E=no-gzip:1]
    RewriteRule \.js\.gz$ - [T=text/javascript,E=no-gzip:1]

    <FilesMatch "(\.js\.gz|\.css\.gz)$">
      # Serve correct encoding type.
      Header set Content-Encoding gzip
      # Force proxies to cache gzipped & non-gzipped css/js files separately.
      Header append Vary Accept-Encoding
    </FilesMatch>
  </IfModule>
</IfModule>
`

**Comment by Graham-72** [comment]
@ColumbiaBlooms Your new redirect needs to be in a new .htaccess file in your surrounding folder i.e C:\xampp\htdocs. It is my practice to use the following content for this .htaccess, where my subdirectory is named backdrop:

```
Options +FollowSymlinks
RewriteEngine on
RewriteRule ^/?$ index.php [L]
RewriteRule index.html?$ index.php [NC,L]
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_URI} !=/favicon.ico
RewriteRule ^(.*)$ backdrop/$1 [L]
```
This causes anything that is not fully addressed with appropriate prefix to a specific file or folder to be redirected to the backdrop folder, with backdrop added into the prefix. In your case you are using bd in place of my use of backdrop.

**Comment by ColumbiaBlooms** [comment]
Thank you, thank you, thank you!!! This works.

**Comment by Graham-72** [comment]
@ColumbiaBlooms I am pleased to hear that your problem is solved. I will attempt to explain what is happening.

Because of the change introduced as a result of #2261 the filter added by CKEditor now provides relative URLs for uploaded files, and this is good for portability. So, depending on where your site is storing files, it provides something like `files/inline-images/Chrysanthemum.jpg` without the leading `/`.

This is then processed in core by function file_create_url($uri) in file.inc which prepends the / and hence your server delivers to your browser /files/inline-images/Chrysanthemum.jpg whereas it needs /bd/files/inline-images/Chrysanthemum.jpg in order to request your image file from the correct location.

The crucial line in the .htaccess file is
`RewriteRule ^(.*)$ bd/$1 [L]`
because this modifies a request for the non-existent file at /files/inline-images/Chrysanthemum.jpg, converting it to a request for /bd/files/inline-images/Chrysanthemum.jpg.

Note that if a request is made for a file that does exist, for example if the initial request from your browser to the server had been for /bd/files/inline-images/Chrysanthemum.jpg, a previous line in the .htacces file would have applied, namely the condition
`RewriteCond %{REQUEST_FILENAME} !-f`
which tests whether such a file exists on your server and only applies the rewrite if the file is not found.

I am thinking that if Backdrop was able to realise that the site is in a subdirectory, function file_create_url($uri) in file.inc could perhaps include this and provide the URL /bd/files/inline-images/Chrysanthemum.jpg in the first place.

**Comment by ColumbiaBlooms** [comment]
@Graham-72 Thanks so much for the explanation. 

I love this kind of support by you folks who are running Backdrop! It is quite refreshing compared to what I got in the larger Drupal community.

**Comment by Graham-72** [comment]
@ColumbiaBlooms Thanks, and you are very welcome. Fortunately I am spared any responsibilty for 'running Backdrop'. I'm just an enthustiastic user and help where I can. Thank you for joining in and contributing. I too very much respect and admire our core team.

**Comment by klonos** [comment]
>I'm just an enthustiastic user and help where I can.

Same boat here.

> I too very much respect and admire our core team.

Second that by a mil.

@Graham-72 great job on providing a solution + a very thorough explanation 👍  and thank you for persisting, providing detailed feedback and trying suggested solutions @ColumbiaBlooms 👍  ...now we need to get this in the docs 😄  (and remember to check if the same problem and solution applies to nginx)

>I am thinking that if Backdrop was able to realise that the site is in a subdirectory, function `file_create_url($uri)` in file.inc could perhaps include this and provide the URL `/bd/files/inline-images/Chrysanthemum.jpg` in the first place.

Yes! Backdrop needs to be "smart" and "make things work" out of the box. Where possible of course.

**Comment by klonos** [comment]
...we also need to check if the problem and solution apply on sites that have their clean URLs either disabled or not available.

**Comment by ColumbiaBlooms** [comment]
> we need to get this in the docs

@klonos Please check https://backdropcms.org/news/backdrop-cms-xampp-windows-10. Scroll down to step 15. Jen and I collaborated on this yesterday to document this fix. There may be other places where this documentation should go as well ... I struggled with Drupal for months with different WAMPs and found that XAMPP works best for me thus the blog.

**Comment by klonos** [comment]
>>we need to get this in the docs

>@klonos Please check https://backdropcms.org/news/backdrop-cms-xampp-windows-10.

Yes, I saw that on Gitter. My comment was before that took place.

> There may be other places where this documentation should go as well...

Yes, I feel that this should go in the official User Guide (once I figure out in which part). The blog post is great, but not the primary place for documentation.

**Comment by klonos** [comment]
...I we fix this issue here by making `file_create_url()` "smarter", we might not need any documentation at all though 😄

**Comment by Graham-72** [comment]
One solution I have found that works on my installations is to use `$_SERVER['PHP_SELF']` to find out whether the installation is in a subdirectory, and if so use the name of the subdirectory as a prefix for the value of `$url` in `core/modules/filter/filter.pages.inc` immediately after the code that has the comment ` // Try to make a local path if possible for better portability.`

This of course results in a loss of portability but it is better than having to manually edit the source code each time an image is added in WYSIWYG.

**Comment by Graham-72** [comment]
This now seems to have been fixed in #2562. Thank you.

**Comment by olafgrabienski** [comment]
@Graham-72 Thanks for documenting the fix in #2562. I think we can close this issue now.

@ all, feel free to reopen if needed.
